File: problems-originals/3101-3200/3135_equalize-strings-by-adding-or-removing-characters-at-ends/solution.cpp
class Solution {
  public:
    int minOperations(string initial, string target) {
        // Characters that survive form a contiguous window of initial and a
        // contiguous window of target, i.e. a common substring; every other
        // character costs exactly one operation, so the answer is
        // m + n - 2 * (longest common substring).
        int best = 0;
        int n = static_cast<int>(target.size());
        vector<int> prev(n + 1, 0);
        for (char a : initial) {
            vector<int> cur(n + 1, 0);
            for (int j = 0; j < n; j++) {
                if (a == target[j]) {
                    cur[j + 1] = prev[j] + 1;
                    best = max(best, cur[j + 1]);
                }
            }
            prev = move(cur);
        }
        return static_cast<int>(initial.size() + target.size()) - 2 * best;
    }
};

File: problems-originals/3101-3200/3135_equalize-strings-by-adding-or-removing-characters-at-ends/solution.rs
impl Solution {
    pub fn min_operations(initial: String, target: String) -> i32 {
        // Characters that survive form a contiguous window of initial and a
        // contiguous window of target, i.e. a common substring; every other
        // character costs exactly one operation, so the answer is
        // m + n - 2 * (longest common substring).
        let t = target.as_bytes();
        let mut best = 0usize;
        let mut prev = vec![0usize; t.len() + 1];
        for &a in initial.as_bytes() {
            let mut cur = vec![0usize; t.len() + 1];
            for (j, &b) in t.iter().enumerate() {
                if a == b {
                    cur[j + 1] = prev[j] + 1;
                    if cur[j + 1] > best {
                        best = cur[j + 1];
                    }
                }
            }
            prev = cur;
        }
        ((initial.len() + target.len()) - 2 * best) as i32
    }
}
